### ISSUE TITLE
500 Internal Server Error

### ISSUE THREAD

**Issue by Ayaan49** [rishabkumar7/devops-qr-code#4]
While testing locally I keep getting this error `Error in generate_qr: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.
INFO:     127.0.0.1:40626 - "POST /generate-qr/?url=https://github.com/rishabkumar7/devops-qr-code/tree/main HTTP/1.1" 500 Internal Server Error` even when all my credentials are correct. I even tested if my credentials were being passed on correctly by running a python script and it worked correctly and proved that I didn't provide wrong AWS credentials. I could only fix this issue when I modified the code to take my credentials explicitly from the `.env` file. So I wanted to know what's the issue and is it okay to modify this code before devopsfying it?

 ```
 from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
import qrcode
import boto3
from io import BytesIO
from dotenv import load_dotenv
import os

# Loading Environment variables
load_dotenv()

app = FastAPI()

# Allowing CORS for local testing
origins = [
    "http://localhost:3000"
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_methods=["*"],
    allow_headers=["*"],
)

# AWS S3 Configuration
aws_access_key_id = os.getenv('AWS_ACCESS_KEY')
aws_secret_access_key = os.getenv('AWS_SECRET_KEY')
aws_region = os.getenv('AWS_DEFAULT_REGION', 'us-west-2')  

s3 = boto3.client('s3',
                  region_name=aws_region,
                  aws_access_key_id=aws_access_key_id,
                  aws_secret_access_key=aws_secret_access_key)

bucket_name = os.getenv('S3_BUCKET_NAME')  

@app.post("/generate-qr/")
async def generate_qr(url: str):
    try:
        # Generate QR Code
        qr = qrcode.QRCode(
            version=1,
            error_correction=qrcode.constants.ERROR_CORRECT_L,
            box_size=10,
            border=4,
        )
        qr.add_data(url)
        qr.make(fit=True)

        img = qr.make_image(fill_color="black", back_color="white")

        # Save QR Code to BytesIO object
        img_byte_arr = BytesIO()
        img.save(img_byte_arr, format='PNG')
        img_byte_arr.seek(0)

        # Generate file name for S3
        file_name = f"qr_codes/{url.split('//')[-1]}.png"

        # Upload to S3
        s3.put_object(Bucket=bucket_name,
                      Key=file_name,
                      Body=img_byte_arr,
                      ContentType='image/png',
                      ACL='public-read')

        # Generate the S3 URL
        s3_url = f"https://{bucket_name}.s3.{aws_region}.amazonaws.com/{file_name}"
        return {"qr_code_url": s3_url}
    except Exception as e:
        print(f"Error in generate_qr: {str(e)}") 
        raise HTTPException(status_code=500, detail=f"An error occurred: {str(e)}")

@app.get("/")
async def root():
    return {"message": "QR Code Generator API is running"}
```

**Comment by rishabkumar7** [comment]
Hi @Ayaan49 thanks for creating this. According the Python file it should load the credentials from the `.env` file, since it use `dotenv` package. Let me test few things and get back to you.

**Comment by Ayaan49** [comment]
That's what I was wondering. I even updated the `dotenv` package to make sure it's working properly. But still it would still give me the same error. 
Yes, please let me know the reason. @rishabkumar7

**Comment by Ayaan49** [comment]
Update: I ran the original code inside the container to see if it was an issue with my machine, since the Docker container isolates the code from my machine, but I encountered the same error. I then used my code with the explicit values inside the container, and it worked. So, is the issue with the `dotenv` package, which may not be loading the correct credentials from the `.env` file, or could it be something else? @rishabkumar7

**Comment by rishabkumar7** [comment]
Thank you @Ayaan49 for the update. I think I have found the issue, will be pushing some fixes.
I had to explicitly use the Secrets in the boto3 initialization.

```
s3 = boto3.client(
    's3',
    aws_access_key_id= os.getenv("AWS_ACCESS_KEY"),
    aws_secret_access_key= os.getenv("AWS_SECRET_KEY"))
```
Based on the [AWS documentation.](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#guide-credentials)

Also, make sure the bucket name is set in the `main.py`.
